File: lib/src/yaml/path_sandbox.dart
import 'dart:io';

import '../core/errors.dart';
import 'diagnostics.dart';
import 'host_config.dart';

/// Validates that [relativePath] resolves inside [config.workspaceRoot].
class PathSandbox {
  /// Creates a sandbox for [config].
  const PathSandbox(this.config);

  /// Host configuration providing workspace root.
  final HostConfig config;

  /// Resolves [relativePath] under the workspace or throws.
  String resolveWorkspaceRelative(String relativePath) {
    final normalized = _normalize(relativePath);
    if (normalized.startsWith('/')) {
      throw PathSandboxException(
        'Absolute paths are not allowed: $relativePath',
        code: 'E_PATH_TRAVERSAL',
      );
    }

    final resolved = _canonical(
      '${config.workspaceRoot.replaceAll('\\', '/')}/$normalized',
    );

    final root = _canonical(config.workspaceRoot.replaceAll('\\', '/'));
    if (!resolved.startsWith(root)) {
      throw PathSandboxException(
        'Path escapes workspace: $relativePath',
        code: 'E_PATH_TRAVERSAL',
      );
    }

    return resolved;
  }

  /// Resolves a template or script path under [config.codemodRoot].
  String resolveTemplateRelative(String relativePath) {
    final combined = '${config.codemodRoot}/$relativePath';
    return resolveWorkspaceRelative(combined);
  }

  /// Resolves a path relative to the codemod root directory.
  String resolveCodemodRelative(String relativePath) {
    final combined = '${config.codemodRoot}/$relativePath';
    return resolveWorkspaceRelative(combined);
  }

  static String _normalize(String path) {
    final segments = path
        .replaceAll('\\', '/')
        .split('/')
        .where((segment) => segment.isNotEmpty && segment != '.')
        .toList();

    if (segments.contains('..')) {
      throw PathSandboxException(
        'Path must not contain "..": $path',
        code: 'E_PATH_TRAVERSAL',
      );
    }

    return segments.join('/');
  }

  static String _canonical(String path) {
    return File(path).absolute.path.replaceAll('\\', '/');
  }
}

/// Converts a [PathSandboxException] to a [RecipeDiagnostic].
RecipeDiagnostic diagnosticFromSandbox(
  PathSandboxException error,
  String file,
) {
  return RecipeDiagnostics.pathSandbox(
    error.message,
    file,
    code: error.code,
  );
}

File: rust/crates/host/src/path_sandbox.rs
use crate::diag_source::source_file_only;
use crate::protocol::RecipeDiagnostic;
use codemod_recipe_core::resource_path::{resolve_under_root, ResourcePathError};
use std::path::PathBuf;

pub type PathSandboxError = ResourcePathError;

/// Validates that relative paths resolve inside the workspace root.
pub struct PathSandbox {
    workspace_root: PathBuf,
}

impl PathSandbox {
    pub fn new(workspace_root: PathBuf) -> Self {
        Self { workspace_root }
    }

    pub fn workspace_root(&self) -> &std::path::Path {
        &self.workspace_root
    }

    pub fn resolve_workspace_relative(
        &self,
        relative_path: &str,
    ) -> Result<PathBuf, PathSandboxError> {
        resolve_under_root(&self.workspace_root, relative_path)
    }

    pub fn resolve_template_relative(
        &self,
        codemod_root: &str,
        relative_path: &str,
    ) -> Result<PathBuf, PathSandboxError> {
        let combined = format!("{codemod_root}/{relative_path}");
        self.resolve_workspace_relative(&combined)
    }
}

pub fn diagnostic_from_sandbox(error: PathSandboxError, file: &str) -> RecipeDiagnostic {
    RecipeDiagnostic::simple(
        "error",
        error.code,
        error.message,
        vec![source_file_only(file)],
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_path_traversal_in_template_paths() {
        let sandbox = PathSandbox::new(std::env::temp_dir());
        let err = sandbox
            .resolve_template_relative(".codemod", "../outside.txt")
            .unwrap_err();
        assert_eq!(err.code, "E_PATH_TRAVERSAL");
    }

    #[test]
    fn resolves_paths_under_workspace() {
        let workspace =
            std::env::temp_dir().join(format!("codemod_sandbox_ok_{}", std::process::id()));
        std::fs::create_dir_all(&workspace).unwrap();
        let sandbox = PathSandbox::new(workspace.clone());
        let resolved = sandbox.resolve_workspace_relative("lib/main.dart").unwrap();
        assert!(resolved.starts_with(&workspace));
        let _ = std::fs::remove_dir_all(workspace);
    }
}
